# Antonio

Group: Final_Project_Group 1 | Workshop: Performance Metrics Classification (MNIST)

Section: Classification, preprocessing, some_digit, precision vs recall cases

Helper class for this section: `MnistViewer` in `mnist_viewer.py`. Helper code lives in the class as methods; this notebook only imports and calls it.

In [1]:
from mnist_viewer import MnistViewer

viewer = MnistViewer()
viewer.load_data()

Loaded 70,000 images with 784 pixels each.


## Antonio

### Classification

**Question:** Summarize in your own words, based on the Wikipedia content and potentially additional sources: What are we trying to solve in a classification problem?

**Answer:**

A classification problem asks which predefined category an observation belongs to, based on its features. In this workshop, the features are the pixel intensities of handwritten digit images. The original task has ten possible classes, from 0 to 9. We simplify it into a binary question: “Does this image represent a 5?” The two classes become “5” and “not 5.”

Sources: [Wikipedia — Statistical classification](https://en.wikipedia.org/wiki/Statistical_classification), [scikit-learn — Getting Started](https://scikit-learn.org/stable/getting_started.html).

**Question:** What is a classifier?

**Answer:**

A classifier is an algorithm or function that maps input features to a class label. In our example, SGDClassifier learns from images and their binary labels. After training, it uses an image’s pixel values to predict True for “5” or False for “not 5.”

Sources: Wikipedia — Statistical classification, scikit-learn — Getting Started.

### Preprocessing

**Question:** What do the Boolean values (`True` and `False`) stand for in the target `y_train_5`?

**Answer:**

The expression y_train == "5" creates a Boolean target for each training image. True means the original label is "5". False means the original label is any other digit. These values are the known training targets, not the model’s predictions.

In [2]:
viewer.prepare_data()


,sample_index,original_label,is_five
0,0,5,True
1,1,0,False
2,2,4,False
3,3,1,False
4,4,9,False


**Question:** Why wasn't `X` changed as well?

**Answer:**

`X` contains the image pixels, which remain useful for deciding whether an image represents a 5. We changed the question we want the model to answer, so we changed the target labels rather than the input images. Each image still has 784 pixel features. The original notebook splits `X` into training and test subsets, but does not change its pixel values when converting the targets to binary labels.

### Training a model

**Question:** Which image does `some_digit` stand for? (scroll above if needed)

**Answer:**

When the original notebook is executed from top to bottom, the last assignment before prediction is some_digit = X[2]. Therefore, it refers to the third image because Python indexing starts at zero. Although an earlier cell assigns X[25001], the later assignment replaces it. Our visualization uses index 2 and confirms that its actual label is "4".

**Visual needed:** Show the some_digit image with its label and the model's prediction.

In [3]:
viewer.train_model()
viewer.plot_digit(index=2).show()

Training complete: the model predicts whether a digit is 5.
Sample index: 2
Actual digit: 4
Actual binary target (is 5): False
Predicted binary target: False (not 5)


**Question:** What is the meaning of the resulting Boolean value (`True` and `False`)?

**Answer:**

A prediction of True means the model classifies the image as a 5. A prediction of False means it classifies the image as not a 5. For X[2], the model returned False, which agrees with its actual label of "4". This is a true negative. The model did not identify the digit specifically as 4; it only rejected the positive class, 5. Because this sample belongs to the training set, this result illustrates prediction but does not establish performance on unseen data.

### Precision vs recall: real-world cases

**Question:** For a medical diagnostics classifier used as a first indicator, what would be better, a high precision or a high recall?

**Answer:**

For the hypothetical first-stage screening system described in the question, I would prioritize high recall. The positive class is “condition present,” and a false negative means missing someone who has the condition and might not be referred for further assessment. High recall reduces the proportion of actual cases missed. Lower precision would create more false alarms and additional assessments, so those costs would still need to be considered.

**Question:** Likewise, what would be better to optimize for when an ML classifier should highlight the best investment opportunities in a stock exchange?

**Answer:**

I would prioritize high precision if the purpose is to produce a selective shortlist of investment opportunities and acting on a false positive is costly. The positive class must first be defined, for example as an investment exceeding a specified return over a fixed period. High precision means a larger proportion of flagged opportunities actually meet that definition. Lower recall means missing some good opportunities. This choice depends on the costs of mistakes; precision alone does not measure profitability or the size of potential losses.

**Question:** Add a real-world case where a high recall is desired, even at the expense of a lower precision.

**Answer:**

A factory could prioritize high recall when screening products for potentially dangerous defects before a human inspection. The positive class is “defective product.” A false negative allows a defective item to pass the screening, while a false positive sends an acceptable item for unnecessary inspection. If missing a dangerous defect is more costly than inspecting an extra item, higher recall is preferable even with lower precision.

**Question:** Add a real-world case where a high precision is preferred, at the expense of a lower recall.

**Answer:**

An email system that automatically deletes messages classified as spam should prioritize high precision for the spam class. A false positive would delete a legitimate message, potentially losing important information. Lower recall would allow some spam into the inbox, but that may be preferable to deleting legitimate messages. The acceptable tradeoff could differ if messages were only placed in a recoverable spam folder.